File: code/python/lists/doubly_linked_list.py
from no_element import NoElement


class DoublyLinkedList:
    """A basic doubly linked list."""

    class Node:
        def __init__(self, item=None, prev=None, nxt=None):
            self.item = item
            self.prev = prev
            self.next = nxt

    def __init__(self):
        self._header = self.Node()
        self._trailer = self.Node(None, self._header, None)  # trailer is preceded by header
        self._header.next = self._trailer  # header is followed by trailer
        self._size = 0

    def size(self):
        return self._size

    def is_empty(self):
        return self._size == 0

    def first(self):
        if self.is_empty():
            return None
        else:
            return self._header.next.item

    def last(self):
        if self.is_empty():
            return None
        else:
            return self._trailer.prev.item

    def _add_between(self, item, predecessor, successor):
        node = self.Node(item, predecessor, successor)
        predecessor.next = node
        successor.prev = node
        self._size += 1

    def add_first(self, item):
        self._add_between(item, self._header, self._header.next)

    def add_last(self, item):
        self._add_between(item, self._trailer.prev, self._trailer)

    def _remove(self, node):
        predecessor = node.prev
        successor = node.next
        predecessor.next = successor
        successor.prev = predecessor
        self._size -= 1

    def remove_first(self):
        if self.is_empty():
            raise NoElement
        self._remove(self._header.next)

    def remove_last(self):
        if self.is_empty():
            raise NoElement
        self._remove(self._trailer.prev)

    def __iter__(self):
        return self.DoublyIterator(self._header.next, self._trailer)

    class DoublyIterator:
        def __init__(self, node, trailer):
            self.walk = node
            self.trailer = trailer

        def __next__(self):
            if self.walk is self.trailer:
                raise StopIteration
            answer = self.walk.item
            self.walk = self.walk.next
            return answer

File: code/java/lists/src/main/java/org/swufe/datastructure/DoublyLinkedList.java
package org.swufe.datastructure;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * A basic doubly linked list.
 */
public class DoublyLinkedList<Item> implements Iterable<Item> {
    private static class Node<Item> {
        Item item;
        Node<Item> prev;
        Node<Item> next;

        public Node(Item item, Node<Item> prev, Node<Item> next) {
            this.item = item;
            this.prev = prev;
            this.next = next;
        }
    }

    private final Node<Item> header;
    private final Node<Item> trailer;
    private int size;

    public DoublyLinkedList() {
        header = new Node<>(null, null, null);
        trailer = new Node<>(null, header, null); // trailer is preceded by header
        header.next = trailer; // header is followed by trailer
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public Item first() {
        if (isEmpty()) return null;
        else return header.next.item;
    }

    public Item last() {
        if (isEmpty()) return null;
        else return trailer.prev.item;
    }

    private void addBetween(Item item, Node<Item> predecessor, Node<Item> successor) {
        Node<Item> node = new Node<>(item, predecessor, successor);
        predecessor.next = node;
        successor.prev = node;
        size++;
    }

    public void addFirst(Item item) {
        addBetween(item, header, header.next);
    }

    public void addLast(Item item) {
        addBetween(item, trailer.prev, trailer);
    }

    private void remove(Node<Item> node) {
        Node<Item> predecessor = node.prev;
        Node<Item> successor = node.next;
        predecessor.next = successor;
        successor.prev = predecessor;
        size--;
    }

    public void removeFirst() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }
        remove(header.next);
    }

    public void removeLast() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }
        remove(trailer.prev);
    }

    @Override
    public Iterator<Item> iterator() {
        return new DoublyIterator();
    }

    private class DoublyIterator implements Iterator<Item> {
        Node<Item> node = header.next;
        @Override
        public boolean hasNext() {
            return node != trailer;
        }

        @Override
        public Item next() {
            Item answer = node.item;
            node = node.next;
            return answer;
        }
    }
}
